### PR TITLE
Recover interrupted eventhubs subscriptions

### PR DESCRIPTION
**Pull request by berndverst**
# Description

This is an attempt to recover eventhubs topic subscription upon error.

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
~* [ ] Created/updated tests~
~* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_~

### CODE REVIEW

File: common/component/azure/eventhubs/eventhubs.go [hunk 1]
```diff
@@ -295,15 +298,29 @@ func (aeh *AzureEventHubs) Subscribe(subscribeCtx context.Context, config Subscr
...
+			} else {
+				if err != nil {
+					aeh.logger.Errorf("Error from event processor: %v", err)
+				} else {
+					aeh.logger.Debugf("Event processor terminated without error")
+				}
+				// wait for subscription loop finished signal
+				<-subscriptionLoopFinished
+				// wait for 5 seconds before restarting the subscription loop
+				<-time.After(5 * time.Second)
```
Comment (ItalyPaleAle) [inline common/component/azure/eventhubs/eventhubs.go:321]: 2 things:

1. Please use a backoff that is exponential (upon success, it should be reset)
2. If the context is canceled, we should stop waiting and return

See how we do here: https://github.dev/dapr/components-contrib/blob/ba6609e624e0404e03d75ccffda6e5de8d94267b/bindings/azure/servicebusqueues/servicebusqueues.go#L145-L155

File: common/component/azure/eventhubs/eventhubs.go [hunk 2]
```diff
@@ -295,15 +298,37 @@ func (aeh *AzureEventHubs) Subscribe(subscribeCtx context.Context, config Subscr
...
-		// Do not log context.Canceled which happens at shutdown
-		if err != nil && !errors.Is(err, context.Canceled) {
-			aeh.logger.Errorf("Error from event processor: %v", err)
+		for {
+			// This is a blocking call that runs until the context is canceled
+			err = processor.Run(subscribeCtx)
+			// Exit if the context is canceled
+			if err != nil && errors.Is(err, context.Canceled) {
+				return
+			} else {
```
Comment (ItalyPaleAle) [inline common/component/azure/eventhubs/eventhubs.go:312]: Nit: remove this `else` block. You already return if there's an error, so we can remove one indentation level and make code easier to read.

File: common/component/azure/eventhubs/eventhubs.go [hunk 3]
```diff
@@ -295,15 +298,37 @@ func (aeh *AzureEventHubs) Subscribe(subscribeCtx context.Context, config Subscr
...
+	go subscriberLoop()
 
 	// Start the processor
 	go func() {
-		// This is a blocking call that runs until the context is canceled
-		err = processor.Run(subscribeCtx)
-		// Do not log context.Canceled which happens at shutdown
-		if err != nil && !errors.Is(err, context.Canceled) {
-			aeh.logger.Errorf("Error from event processor: %v", err)
+		for {
```
Comment (ItalyPaleAle) [inline common/component/azure/eventhubs/eventhubs.go:306]:
```suggestion

	// Start the processor
	go func() {
		for {
	        go subscriberLoop()

```

File: common/component/azure/eventhubs/eventhubs.go [hunk 4]
```diff
@@ -295,15 +298,37 @@ func (aeh *AzureEventHubs) Subscribe(subscribeCtx context.Context, config Subscr
...
+				select {
+				case <-subscribeCtx.Done():
+					return
+				case <-subscriptionLoopFinished:
+					// Waiting here is not strictly necessary, however, we will wait for a short time to increase the likelihood of transient errors having disappeared
+					select {
+					case <-subscribeCtx.Done():
+						return
+					case <-time.After(5 * time.Second):
+						go subscriberLoop()
```
Comment (ItalyPaleAle) [inline common/component/azure/eventhubs/eventhubs.go:328]:
```suggestion
						// Nop, restart the loop
```

File: common/component/azure/eventhubs/eventhubs.go [hunk 5]
```diff
@@ -295,15 +298,37 @@ func (aeh *AzureEventHubs) Subscribe(subscribeCtx context.Context, config Subscr
...
+				if err != nil {
+					aeh.logger.Errorf("Error from event processor: %v", err)
+				} else {
+					aeh.logger.Debugf("Event processor terminated without error")
+				}
+				// wait for subscription loop finished signal
+				select {
+				case <-subscribeCtx.Done():
+					return
+				case <-subscriptionLoopFinished:
```
Comment (ItalyPaleAle) [inline common/component/azure/eventhubs/eventhubs.go:322]: Nit: here too, I'd create a secondary select *after* to make code more readable, so there's no nested select's

**Review by ItalyPaleAle — COMMENTED** [review]
Thanks, just some comments for code readability

File: common/component/azure/eventhubs/eventhubs.go (hunk 3 above)
Comment (ItalyPaleAle) [reply common/component/azure/eventhubs/eventhubs.go:306]: Whoops this came out wrong. The idea was just to move "go subscrierLoop" inside the for loop, so then at the end you don't need to invoke it again. Just for code readability - makes it easier to follow the flow

Comment (berndverst) [reply common/component/azure/eventhubs/eventhubs.go:306]: I don't find that easier to read but as you wish.

File: common/component/azure/eventhubs/eventhubs.go (hunk 5 above)
Comment (berndverst) [reply common/component/azure/eventhubs/eventhubs.go:322]: I'm not sure what you mean by that.

File: common/component/azure/eventhubs/eventhubs.go (hunk 3 above)
Comment (berndverst) [reply common/component/azure/eventhubs/eventhubs.go:306]: Ah I see what you mean. I thought you wanted me to inline the entire function - which would have been messy. I agree with what you are saying

File: common/component/azure/eventhubs/eventhubs.go (hunk 4 above)
Comment (berndverst) [reply common/component/azure/eventhubs/eventhubs.go:328]: This comment only makes sense based on your other suggestion ;)

**Review by ItalyPaleAle — APPROVED** [review]

**Comment by yaron2** [comment]
> # Components certification test
> 🔗 **[Link to Action run](https://github.com/dapr/components-contrib/actions/runs/7835847682)**
> 
> Commit ref: [bb2e600](https://github.com/dapr/components-contrib/commit/bb2e6001f51e0767ae56ffaf1f94e44731deb04b)
> 
> # ❌ Some certification tests failed
> These tests failed:
> 
> * bindings.azure.eventhubs
> * bindings.rabbitmq
> * pubsub.aws.snssqs
> * pubsub.mqtt3
> * pubsub.pulsar

Tests failed here, including the eventhubs one.

**Comment by berndverst** [comment]
> > # Components certification test
> > 🔗 **[Link to Action run](https://github.com/dapr/components-contrib/actions/runs/7835847682)**
> > Commit ref: [bb2e600](https://github.com/dapr/components-contrib/commit/bb2e6001f51e0767ae56ffaf1f94e44731deb04b)
> > # ❌ Some certification tests failed
> > These tests failed:
> > 
> > * bindings.azure.eventhubs
> > * bindings.rabbitmq
> > * pubsub.aws.snssqs
> > * pubsub.mqtt3
> > * pubsub.pulsar
> 
> Tests failed here, including the eventhubs one.

@yaron2 the eventhubs bindings test passed in this PR before - problem is that the 4  conformance and certification tests for eventhubs can't be run in parallel - they weren't designed well, so you manually need to restart them.

**Pull request merged**